kimchi/src/circuits/polynomials/foreign_field_add/gadget.rs

```diff
@@ -0,0 +1,414 @@
+use std::collections::HashMap;
+
+use ark_ff::{FftField, SquareRootField, Zero};
+use ark_poly::{
+    univariate::DensePolynomial, EvaluationDomain, Evaluations, Radix2EvaluationDomain as D,
+};
+use array_init::array_init;
+use rand::{prelude::StdRng, SeedableRng};
+
+use crate::{
+    alphas::Alphas,
+    circuits::{
+        argument::{Argument, ArgumentType},
+        constraints::ConstraintSystem,
+        expr::{self, l0_1, Environment, LookupEnvironment, E},
+        gate::{CircuitGate, CircuitGateError, CircuitGateResult, GateType},
+        lookup::{
+            self,
+            lookups::{LookupInfo, LookupsUsed},
+            tables::{GateLookupTable, LookupTable},
+        },
+        polynomial::COLUMNS,
+        wires::{GateWires, Wire},
+    },
+};
+
+use super::FFAdd;
+
+impl<F: FftField + SquareRootField> CircuitGate<F> {
+    /// Create foreign field addition gate
+    ///     Inputs the starting row
+    ///     Outputs tuple (next_row, circuit_gates) where
+    ///       next_row      - next row after this gate
+    ///       circuit_gates - vector of circuit gates comprising this gate
+    pub fn create_foreign_field_add(start_row: usize) -> (usize, Vec<Self>) {
+        // Create multi-range-check gates for $a, b, r, s$
```

Comment:
Absolutely agree with you in the optimization.  When these terms are private input, this proof of correctness should be provided, but I agree that once is enough. You don't want to check this range many times for the same elements basically. This would be specially relevant in case we were chaining these operations, in whose case we will need an "augmentable" version of the gadget to only include checks for new terms once. Now, it remains to decide what would happen if the terms come from public input. Should we just assume they have the right shape as the verifier could be able to check so? Or we may want to be more conservative?

Reply:
oh yeah I guess since it's a gadget it doesn't matter :D we can use a different gadget later in snarky if we deem that it's not reasonable. Maybe good enough for now to document that these checks might not be useful and are over precaucious